presto-main-base/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java

```diff
@@ -1465,6 +1465,100 @@ public void testArrayL2SquaredDouble()
                 DOUBLE, null);
     }
 
+    @Test
+    public void testArrayDotProduct()
+    {
+        // functionality test
+        assertFunction(
+                "dot_product(array[DOUBLE '1.0', DOUBLE '2.0', DOUBLE '3.0'], array[DOUBLE '1.0', DOUBLE '2.0', DOUBLE '3.0'])",
+                DOUBLE, 14.0d);
+        assertFunction(
+                "dot_product(array[DOUBLE '1.0', DOUBLE '2.0', DOUBLE '3.0'], array[DOUBLE '4.0', DOUBLE '5.0', DOUBLE '6.0'])",
+                DOUBLE, 32.0d);
+        assertFunction(
+                "dot_product(array[DOUBLE '-1.0', DOUBLE '-2.0', DOUBLE '-3.0'], array[DOUBLE '1.0', DOUBLE '2.0', DOUBLE '3.0'])",
+                DOUBLE, -14.0d);
+        assertFunction(
+                "dot_product(array[DOUBLE '0.0', DOUBLE '0.0', DOUBLE '0.0'], array[DOUBLE '0.0', DOUBLE '0.0', DOUBLE '0.0'])",
+                DOUBLE, 0.0d);
+        // identical size test
+        assertInvalidFunction(
+                "dot_product(array[DOUBLE '1.0', DOUBLE '2.0'], array[DOUBLE '1.0', DOUBLE '2.0', DOUBLE '3.0'])",
+                "Both array arguments must have identical sizes");
+        // null test
+        assertFunction(
+                "dot_product(CAST(null AS array(double)), array[DOUBLE '1.0', DOUBLE '2.0', DOUBLE '3.0'])",
+                DOUBLE, null);
+        assertFunction(
+                "dot_product(array[DOUBLE '1.0', DOUBLE '2.0', DOUBLE '3.0'], CAST(null AS array(double)))",
+                DOUBLE, null);
+        assertFunction(
+                "dot_product(CAST(null AS array(double)), CAST(null AS array(double)))",
+                DOUBLE, null);
+        // treat null inside array as zero
```

Comment:
It's not clear to me that this should be the behavior.  SQL scalar product and addition both propagate nulls.  Can you explain your reasoning on treating NULL as 0?